src/WinUsb/WinUsb.Helpers.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright © .NET Foundation and Contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace PInvoke
+{
+    using System;
+    using System.Runtime.InteropServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <content>
+    /// Methods and nested types that are not strictly P/Invokes but provide
+    /// a slightly higher level of functionality to ease calling into native code.
+    /// </content>
+    public static partial class WinUsb
+    {
+        /// <summary>
+        /// Asynchronously reads data from the specified pipe.
+        /// </summary>
+        /// <param name="interfaceHandle">
+        /// An opaque handle to the interface that contains the endpoint with which the pipe is associated.
+        /// </param>
+        /// <param name="pipeID">
+        /// <paramref name="pipeID"/> corresponds to the bEndpointAddress field in the endpoint descriptor.
+        /// </param>
+        /// <param name="buffer">
+        /// A caller-allocated buffer that receives the data that is read.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// A <see cref="CancellationToken"/> which can be used to cancel the asynchronous operation.
+        /// </param>
+        /// <returns>
+        /// A <see cref="ValueTask"/> which represents the asynchronous operation, and returns the number of bytes transferred.
+        /// </returns>
+        public static unsafe ValueTask<int> WinUsb_ReadPipeAsync(SafeUsbHandle interfaceHandle, byte pipeID, Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            var overlapped = new WinUsbOverlapped(interfaceHandle, pipeID, buffer);
```

Comment:
Two things came up when trying to avoid allocating the `WinUsbOverlapped` class here:
- The native API takes an `OVERLAPPED*` pointer to a struct. You'll need a place to store that struct and pin it. `Overlapped.Pack` does that for you.
- You'll need to set `NativeOverlapped.EventHandle `. The sample code I could find online uses `ManualResetEvent.SafeWaitHandle.DangerousGetHandle`. But `ManualResetEvent` does not provide a way to asynchronously await that event (there's only `WaitOne`, which would block). Plus, `ManualResetEvent` is a `class`.

One way to avoid allocating a class _on every read/write call_ may be to implement a higher-level class, like `WinUsbStream`, and cache `WinUsbOverlapped` in that class.

Reply:
You *can* actually *await* a `WaitHandle` (which `ManualResetEvent` derives from) using [this extension method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.awaitextensions.getawaiter?view=visualstudiosdk-2019#Microsoft_VisualStudio_Threading_AwaitExtensions_GetAwaiter_System_Threading_WaitHandle_). I don't think we'd want to add a dependency on that library, but you could do the same [thing that library does](https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/TplExtensions.cs#L457) using just .NET APIs. It's certainly complicated though, and involves allocations of its own.
I'm sharing the above for your info, but I am satisfied with your research and your current approach. Thanks for investigating it.

Reply:
Thanks, I created #514 to track support of asynchronously awaiting `WaitHandle`s, and #515 to discuss the idea of implementing a `WinUsbPipeStream` as a helper.